File: c++/inoneshort/oops/polymorphism.cpp
#include <iostream>
using namespace std;
// function overloading
class sum {
    public:
        void add(int x, int y){
            int sum = x+y;
            cout<<sum<<endl;
        }
        void add(int x, int y , int z){
            int sum = x+y+z;
            cout<<sum<<endl;
        }
        void add(float x, float y){
            float sum = x+y;
            cout<<sum<<endl;
        }


};
// operator overloading

int main(){
    // FUNCTION OVERLOADING 
    sum s;
    s.add(3,5);
    s.add(2,6,4);
    s.add(float(2.4) , float(5.9));

    return 0;
}
/*
 Polymorphism -->ability of object / method to take different forms
  TYPES 
  1. compile time polymorphism
  2. run time polymorphism
  function overloading --> define a number of function  with same function name they perform differently according to the arrgument passed 

*/

File: python/collegewallah/ch 7 oop/polymorphism.py
# polymorphism 
# allow object of different classes to behave as object of thr same super class 

# compile time polymorphism

#method overloading 
class add:
    def sum(self, a,b):
        print (a+b)

    def sum(self, a,b,c):
        print(a+b+c)

a1=add()
a1.sum(8,9,5)
a1.sum(8,9,11)
# OPERATOR OVER LOADING 

class complexnumber:
    def __init__(self, real,img):
        self.real=real
        self.img=img

    def __add__(self,other):
        return complexnumber(self.real+other.real, self.img+ other.img)
c1= complexnumber(1,2)
c2= complexnumber(3,4)
c3= c1+c2
print(c3.real, " + i",c3.img)